### ISSUE TITLE
possible python error at "Correcting umis" step

### ISSUE THREAD

**Issue by sorjuela** [Hoohm/CITE-seq-Count#110]
Hello, I have data from TotalSeqA. I received FASTQ files that were obtained with `cellranger mkfastq`, and concatenated all the R1 and R2 files into one to run Cite-seq-Count:
```
cat ADT1_S1_L00*_R1_001.fastq.gz HTO1_S1_L00*_R1_001.fastq.gz > ADT_HTO_1_S1_R1_001.fastq.gz
cat ADT1_S1_L00*_R2_001.fastq.gz HTO1_S1_L00*_R2_001.fastq.gz > ADT_HTO_1_S1_R2_001.fastq.gz
```
As you can see, I have ADTs and HTOs separated, and 2 lanes per sample. 

I installed CITE-seq-Count within a conda environment using pip-install, using 
```
conda config --set pip_interop_enabled True
pip install CITE-seq-Count==1.4.3
```
I get the error at the "Correcting umis" step (see attached log file). The error message is not very specific, so I don't really know how to troubleshoot. Can you help me out? Attached is also the tags.csv (I loaded a .txt, but i ran the .csv)
[tags.txt](https://github.com/Hoohm/CITE-seq-Count/files/4159556/tags.txt)
[cite-seq-count.log](https://github.com/Hoohm/CITE-seq-Count/files/4159543/cite-seq-count.log)

Versions:
conda version : 4.7.12
python version : 3.7.4.final.0
Linux/4.4.0-127-generic ubuntu/16.04.6 glibc/2.23

**Comment by Hoohm** [comment]
Hello @sorjuela,
this is kind of concerning as I remember fixing this bug a while ago.
I don't have much time to work on it these days so it might take a while to figure out what's happening as it's also the first time I see it creep up.

Just to be sure, could you run `CITE-seq-Count --version` for me please.

**Comment by sorjuela** [comment]
Hi @Hoohm, yes output is `CITE-seq-Count v1.4.3`

**Comment by EnoraBioinfo** [comment]
Hello, I have exactly the same error.

I get the error after "Correcting umis" :
  File "/sandbox/users/egaschet/miniconda3/envs/mynewenv/lib/python3.8/site-packages/cite_seq_count/processing.py", line 229, in correct_umis
    for TAG in final_results[cell_barcode]:
RuntimeError: dictionary keys changed during iteration

@sorjuela did you find out what the problem was ?

**Comment by sorjuela** [comment]
nope, was hoping the author would give me some suggestions/alternatives on how to continue. I'm not using CITE-seq-Count for the moment.

**Comment by Hoohm** [comment]
Hello @sorjuela and @EnoraBioinfo I'm sorry for the long waiting time. I'm swamped and only working on CITE-seq-Count on the weekends when I don't have other projects to do.

I want to try and debug this.

Do you have any duplicated tags in your tags.csv files?

**Comment by sorjuela** [comment]
Hi @Hoohm, no worries. I don't see any duplicates in mine. You can find my tags file in my first comment above.

**Comment by EnoraBioinfo** [comment]
I don't see any duplicated tags either :
TGTCTTTCCTGCCAG,antihuman_H7
CTCCTCTGCAATTAC,antihuman_H8
TAACGACCAGCCATA,antihuman_H12
AAATCTCTCAGGCTC,antihuman_H13
CTGTATGTCCGATTG,antihuman_H14
TAAGATTCAGAGCGA,antihuman_H15

**Comment by Hoohm** [comment]
I'm having a hard time recreating the error.

Would it be possible to send me some data?

**Comment by Hoohm** [comment]
Best case scenario would a small subset that still crashes. You can use the `-n` option to try it out.

**Comment by Hoohm** [comment]
Do you guys provide a whitelist when you run CSC?

**Comment by sorjuela** [comment]
Hi @Hoohm sorry for the delay. I can send you some data privately, i will send it to your gmail if that works for you. About the whitelist I'm not providing it, I thought the tags.csv would be enough?

**Comment by sorjuela** [comment]
also, I uploaded the log file in my first comment, so you can see how I'm running the software.

**Comment by Hoohm** [comment]
@sorjuela by email would be perfect yes.

I was asking about the whitelist because I was suspecting some issues with missing barcodes that would be added and could be responsible for the bug.

As you mentioned, the call is in the log and since you did not provide a whitelist, this is not the problem.

If you send the data, I can run and debug it very easily. That would be amazing.

**Comment by sorjuela** [comment]
Ok, I sent you a link to download the files.

**Comment by ligrob** [comment]
Hi all, 

have you by chance already found a solution for this? Unfortunately, I stumbled across the same error as mentioned above. Any advice on how to handle this would be great!

**Comment by Hoohm** [comment]
Hello @ligrob 
This is an old bug. We fixed it by using the latest release, 1.4.3.

**Comment by robinfchan** [comment]
Yeah unfortunately I'm also still getting this problem even when running 1.4.3 :(
[log-events-viewer-result.txt](https://github.com/Hoohm/CITE-seq-Count/files/4806162/log-events-viewer-result.txt)

**Comment by Hoohm** [comment]
Are you using this locally on a clean install?
If you are, would it be possible to send me a sample on which it crashes so that I can try it as well?

**Comment by codymowery** [comment]
@Hoohm what is your email so I can send a reproducible example? I am having the same issue running version=1.4.3. I am currently running with `--no_umi_correction` to see if that prevents the issue, but that solution is less than ideal. Thanks!

**Comment by robinfchan** [comment]
@codymowery check out https://github.com/Hoohm/CITE-seq-Count/issues/125

try running with python 3.7

**Comment by codymowery** [comment]
Looks like that did the trick. Thanks @robinfchan!

**Comment by christ-lili** [comment]
@robinfchan  Thanks Robinchan, py37 is working but py38 will produce the bug.

**Comment by Hoohm** [comment]
It should be fixed in 1.4.4 that just came out
